### PR TITLE
JSONFileHandler - lebmLvaS

### PR DESCRIPTION
**Pull request by collinnolen**
Created JSONFileHanlder class which will handle retrieving and saving JSON files. Also added the AchievmentHanlder class which will handle the player's achievement tracking (just tracking some stats at the moment). 

SingleRun and LifeTime classes will represent the structure of the JSON format. By doing it this way we can simple serialize and deserialize the C# object to JSON string incredibly easily. 

**Update 1:**
Created StreamingAssets folder, Google 'Unity StreamingAssets folder'  for more info about this cool folder, will be a great place for JSON files. 

Also added the core of JSONFileHandler logic, handles finding the JSON file, reading in the text, and returning an object whose type matches the one you pass into the function. If no file exists, it will create a new instance of the object type that you passed in.

 When saving, it will overwrite the JSON file if it already exists, but this is okay because we are keeping track of all the data in memory when we deserialize the JSON object as a C# object.

**Update 2:**
Added IObserver and ISubject interfaces to facility the Observer pattern for the AchievementHandler class. I implemented just a very simple achievement tracking system (just tracks ships slain and gold collected, never acts upon it). To save the current stats JSON use the Num Lock key.

### CODE REVIEW

**Comment by collinnolen** [comment]
Work in progress.

**Comment by collinnolen** [comment]
Created StreamingAssets folder, Google 'Unity StreamingAssets folder'  for more info about this cool folder, will be a great place for JSON files. 

Also added the core of JSONFileHandler logic, handles finding the JSON file, reading in the text, and returning an object whose type matches the one you pass into the function. If no file exists, it will create a new instance of the object type that you passed in.

 When saving, it will overwrite the JSON file if it already exists, but this is okay because we are keeping track of all the data in memory when we deserialize the JSON object as a C# object.

**Comment by artsiriwatt** [comment]
- extracting subject logic to a parent class
- fixing enum - remove casting

**Comment by collinnolen** [comment]
> * [x]  extracting subject logic to a parent class
> * [x]  fixing enum - remove casting
> * [x]  hack in - add gold when killing a ship for testing purposes

Completed

**Comment by collinnolen** [comment]
> * [x]  extracting subject logic to a parent class
> * [x]  fixing enum - remove casting
> * [x]  hack in - add gold when killing a ship for testing purposes

duplicate

> Looks great. I like the use of C# classes to help serialization of JSONs. Also StreamingAssets will be super handy. All I noticed was a function misspell in SingleRun.cs, the "AddCurrentColdCollected" should be "AddCurrentGoldCollected". Nice work! Good to merge after fix.

Fixed.

**Comment by collinnolen** [comment]
Made the changes, please review again and Merge if changes are ready.

File: Assets/Scripts/Navigation/NavController.cs [hunk 1]
```diff
@@ -1,21 +1,32 @@
...
+using Assets.Scripts.shared.Achievements;
+using Assets.Scripts.shared;
+using Assets.Scripts.shared.ObserverPattern;
 
-public class NavController : MonoBehaviour {
+public class NavController : SubjectMonobehavior
+{
 
     public static NavController instance = null;
+    public static AchievementHandler achievementHandlerInstance = null;
```
Comment (artsiriwatt) [inline Assets/Scripts/Navigation/NavController.cs]: Can this move into SubjectMonobehavior

File: Assets/Scripts/Navigation/NavController.cs [hunk 2]
```diff
@@ -1,21 +1,32 @@
...
     public Map map;
 
-    private static bool backFromCombat;
 
     //private int level = 1;
 
     void Awake()
     {
+        //Create Singleton of AchievementHandler and add it to observer list.
+        if(achievementHandlerInstance == null)
```
Comment (artsiriwatt) [inline Assets/Scripts/Navigation/NavController.cs]: Also this part, and then call super.Awake() in the method

File: Assets/Scripts/Navigation/NavController.cs [hunk 3]
```diff
@@ -64,6 +75,7 @@ public void CreatePlayer()
     {
         //Temp Code for development
         StaticPlayer.Player = new PlayerData(1000, 2000, 1000, 2000, .1f, 10, 4, 3, 10);
+        StaticPlayer.Player.AddObserver(achievementHandlerInstance);
```
Comment (artsiriwatt) [inline Assets/Scripts/Navigation/NavController.cs]: Same here

**Comment by artsiriwatt** [comment]
Plan is address to resolve previous comments in this PR to a refactor ticket in future

**Pull request merged**